Nama: Bernard Hugo

NIM: 2540124450

Kelas: LA05

UAS Deep Learning

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

**Bagian a**

In [ ]:
df = pd.read_csv('drive/MyDrive/Emotion.csv')
print(df)

                                                    text    label
0                                i didnt feel humiliated  sadness
1      i can go from feeling so hopeless to so damned...  sadness
2       im grabbing a minute to post i feel greedy wrong    anger
3      i am ever feeling nostalgic about the fireplac...     love
4                                   i am feeling grouchy    anger
...                                                  ...      ...
19995  im having ssa examination tomorrow in the morn...  sadness
19996  i constantly worry about their fight against n...      joy
19997  i feel its important to share this info for th...      joy
19998  i truly feel that if you are passionate enough...      joy
19999  i feel like i just wanna buy any cute make up ...      joy

[20000 rows x 2 columns]


In [ ]:
df.isnull().sum()

text     0
label    0
dtype: int64

Melakukan pengecekan pada dataset apakah terdapat data-data yang kosong pada dataset. Hasilnya tidak ada data yang kosong didalam dataset tersebut.

In [ ]:
label = df["label"].unique()

for i in label:
    print(i)

sadness
anger
love
surprise
fear
joy


Melihat kelas-kelas yang ada pada atribut "label". Berguna juga untuk mengetahui berapa jumlah kelas yang ada pada atribut "label". Kelas-kelas yang ada pada "label" adalah sadness, anger, love, surprise, fear, dan joy.

Melakukan preprocessing

In [ ]:
x = df['text'].values
y = df['label'].values

In [ ]:
label = {
    'sadness': 0,
    'anger': 1,
    'love': 2,
    'surprise': 3,
    'fear': 4,
    'joy': 5
}

encoded_y = [label[i] for i in df["label"]]

print(encoded_y)

[0, 0, 1, 2, 1, 0, 3, 4, 5, 2, 0, 5, 1, 0, 5, 5, 0, 0, 0, 4, 1, 4, 5, 5, 1, 0, 0, 0, 1, 5, 5, 4, 3, 1, 5, 5, 5, 5, 1, 5, 5, 5, 5, 5, 0, 0, 5, 2, 5, 1, 5, 0, 1, 4, 5, 0, 0, 3, 5, 5, 5, 2, 4, 4, 3, 1, 1, 0, 2, 5, 0, 0, 5, 0, 0, 0, 5, 5, 5, 1, 0, 1, 1, 1, 5, 5, 5, 5, 0, 4, 2, 1, 0, 1, 2, 0, 5, 5, 0, 1, 2, 5, 5, 5, 0, 5, 5, 5, 5, 0, 5, 5, 2, 0, 0, 5, 5, 0, 5, 5, 4, 4, 4, 0, 2, 5, 5, 2, 4, 3, 5, 5, 5, 5, 1, 4, 5, 1, 2, 1, 0, 5, 0, 1, 5, 3, 0, 1, 1, 0, 5, 4, 5, 5, 4, 0, 3, 3, 5, 1, 4, 1, 0, 1, 0, 4, 0, 5, 3, 4, 5, 1, 5, 1, 5, 4, 4, 0, 5, 5, 0, 3, 5, 4, 5, 0, 5, 5, 5, 4, 5, 3, 5, 1, 0, 0, 5, 1, 0, 5, 5, 3, 5, 4, 5, 4, 0, 4, 2, 0, 4, 2, 0, 0, 1, 5, 2, 1, 0, 3, 1, 5, 0, 1, 5, 5, 5, 5, 5, 0, 5, 2, 5, 0, 1, 3, 5, 1, 5, 2, 0, 0, 5, 0, 0, 5, 0, 5, 5, 5, 4, 0, 1, 0, 1, 2, 5, 2, 0, 5, 5, 5, 0, 5, 0, 1, 2, 0, 2, 0, 0, 0, 0, 0, 0, 4, 5, 0, 0, 5, 2, 0, 1, 0, 2, 5, 5, 5, 0, 5, 0, 5, 5, 5, 0, 0, 0, 2, 4, 0, 4, 5, 5, 4, 5, 1, 1, 2, 0, 3, 5, 1, 0, 0, 1, 2, 3, 0, 2, 5, 1, 5, 0, 0, 5, 5, 4, 0, 1, 5, 2, 5, 5, 

Melakukan preprocessing untuk mengubah kategori kelas-kelas yang ada pada atribut "label" dari string menjadi numerik.

**Bagian b**

**Membagi dataset menjadi training, validation, dan test (70:15:15)**

In [ ]:
train_size = int(len(x) * 0.7)
test_size = int(len(x) * 0.15)
val_size = len(x) - (train_size - test_size)

x_train, y_train = x[:train_size], encoded_y[:train_size]
x_test, y_test = x[train_size:train_size + test_size], encoded_y[train_size:train_size + test_size]
x_val, y_val = x[train_size + test_size:], encoded_y[train_size + test_size:]

**Bagian c**

**Membuat DistilBERT model**

In [ ]:
pip install --upgrade keras-nlp

     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 415.4/415.4 kB 5.6 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 950.8/950.8 kB 10.8 MB/s eta 0:00:00
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.2/5.2 MB 20.7 MB/s eta 0:00:00


In [ ]:
import keras_nlp

preprocessor = keras_nlp.models.DistilBertPreprocessor.from_preset(
    "distil_bert_base_en_uncased",
    sequence_length = 130
)

model = keras_nlp.models.DistilBertClassifier.from_preset(
    "distil_bert_base_en_uncased",
    num_classes = 6,
    preprocessor = preprocessor
)

Using TensorFlow backend


100%|██████████| 580/580 [00:00<00:00, 1.35MB/s]
100%|██████████| 226k/226k [00:00<00:00, 61.8MB/s]
100%|██████████| 515/515 [00:00<00:00, 291kB/s]
100%|██████████| 253M/253M [00:02<00:00, 93.0MB/s]
/usr/local/lib/python3.10/dist-packages/keras_nlp/src/models/backbone.py:37: UserWarning: `Model.state_updates` will be removed in a future version. This property should not be used in TensorFlow 2.0, as `updates` are applied automatically.
  return id(getattr(self, attr)) not in self._functional_layer_ids
/usr/local/lib/python3.10/dist-packages/keras_nlp/src/models/backbone.py:37: UserWarning: `layer.updates` will be removed in a future version. This property should not be used in TensorFlow 2.0, as `updates` are applied automatically.
  return id(getattr(self, attr)) not in self._functional_layer_ids


In [ ]:
model.fit(x = x_train, y = y_train, validation_data = (x_val, y_val), batch_size = 32, epochs = 5)

Epoch 1/5
438/438 [==============================] - 266s 458ms/step - loss: 0.4998 - sparse_categorical_accuracy: 0.8239 - val_loss: 0.1804 - val_sparse_categorical_accuracy: 0.9243
Epoch 2/5
438/438 [==============================] - 189s 432ms/step - loss: 0.1471 - sparse_categorical_accuracy: 0.9376 - val_loss: 0.1577 - val_sparse_categorical_accuracy: 0.9320
Epoch 3/5
438/438 [==============================] - 188s 428ms/step - loss: 0.1138 - sparse_categorical_accuracy: 0.9484 - val_loss: 0.1588 - val_sparse_categorical_accuracy: 0.9307
Epoch 4/5
438/438 [==============================] - 216s 494ms/step - loss: 0.0887 - sparse_categorical_accuracy: 0.9571 - val_loss: 0.1615 - val_sparse_categorical_accuracy: 0.9360
Epoch 5/5
438/438 [==============================] - 187s 427ms/step - loss: 0.0742 - sparse_categorical_accuracy: 0.9664 - val_loss: 0.2091 - val_sparse_categorical_accuracy: 0.9257


**Bagian d**

**Menemukan skor precision, recall, F1-Score dan accuracy**

In [ ]:
from sklearn.metrics import classification_report
from sklearn.metrics import precision_score, recall_score, f1_score, accuracy_score

y_pred = model.predict(x_test)
y_pred = np.argmax(y_pred, axis = 1)

print(classification_report(y_test, y_pred))

94/94 [==============================] - 19s 169ms/step
              precision    recall  f1-score   support

           0       0.95      0.99      0.97       910
           1       0.93      0.93      0.93       418
           2       0.80      0.94      0.86       249
           3       0.93      0.71      0.81       100
           4       0.86      0.93      0.90       364
           5       0.98      0.91      0.94       959

    accuracy                           0.93      3000
   macro avg       0.91      0.90      0.90      3000
weighted avg       0.94      0.93      0.93      3000



In [ ]:
print(f"Prescision:  {precision_score(y_test, y_pred, average = 'macro') :.3f}")
print(f"Recall: {recall_score(y_test, y_pred, average = 'macro') :.3f}")
print(f"F1-Score: {f1_score(y_test, y_pred, average = 'macro') :.3f}")
print(f"Accuracy: {accuracy_score(y_test, y_pred) :.3f}")

Prescision:  0.912
Recall: 0.899
F1-Score: 0.902
Accuracy: 0.932


Dari hasil precision, recall, F1-Score dan accuracy sangat tinggi yang berarti model DistilBERT yang dibuat melakukan kinerja yang baik dalam melakukan klasifikasi. Nilai precision yang dihasilkan sebesar 0.912 (91.2%), recall sebesar 0.899 (89.9%), F1-Score sebesar 0.902 (90.2%), dan accuracy sebesar 0.932 (93.2%).

Nilai precision yang dihasilkan menunjukkan bahwa perolehan nilai yang benar-benar positif yang dilakukan oleh model DistilBERT sangat baik. Berdasarkan nilai recall yang dihasilkan, menunjukkan bahwa model banyak menemukan nilai yang terprediksi positif dengan baik. Nilai F1-Score yang tinggi menunjukkan hubungan precision dan recall sangat seimbang. Nilai accuracy menjadi nilai yang berperan penting dalam mengukur kinerja model dalam melakukan tugas. Sehingga dari nilai accuracy tersebut, model melakukan kinerja yang bagus dalam melakukan klasifikasi.